submission/distributor_test.go

```diff
@@ -102,10 +120,94 @@ func TestNewDistributorLogClients(t *testing.T) {
 		t.Run(tc.name, func(t *testing.T) {
 			_, err := NewDistributor(tc.ll, ctpolicy.ChromeCTPolicy{}, tc.lcBuilder)
 			if gotErr, wantErr := err != nil, tc.errRegexp != nil; gotErr != wantErr {
-				t.Errorf("Expected error state does not match produced one.")
+				var unwantedErr string
+				if gotErr {
+					unwantedErr = fmt.Sprintf(" %q ", err)
+				}
+				t.Errorf("Got error = %v%s, expected error = %v", gotErr, unwantedErr, wantErr)
 			} else if tc.errRegexp != nil && !tc.errRegexp.MatchString(err.Error()) {
 				t.Errorf("Error %q did not match expected regexp %q", err, tc.errRegexp)
 			}
 		})
 	}
 }
+
+// Stub for AddLogCLient interface
+type stubLogClient struct {
+	logURL string
+}
+
+func (m stubLogClient) AddChain(ctx context.Context, chain []ct.ASN1Cert) (*ct.SignedCertificateTimestamp, error) {
+	return nil, nil
+}
+
+func (m stubLogClient) AddPreChain(ctx context.Context, chain []ct.ASN1Cert) (*ct.SignedCertificateTimestamp, error) {
+	return nil, nil
+}
+
+func (m stubLogClient) GetAcceptedRoots(ctx context.Context) ([]ct.ASN1Cert, error) {
+	roots := []ct.ASN1Cert{}
+	if certs, ok := RootsCerts[m.logURL]; ok {
+		for _, cert64 := range certs {
+			cert, err := base64.StdEncoding.DecodeString(cert64)
+			if err != nil {
+				return nil, err
+			}
+			roots = append(roots, ct.ASN1Cert{Data: cert})
+		}
+	}
+	return roots, nil
+}
+
+func buildstubLogClient(log *loglist.Log) (client.AddLogClient, error) {
+	return stubLogClient{logURL: log.URL}, nil
+}
+
+func TestNewDistributorRootPools(t *testing.T) {
+	testCases := []struct {
+		name         string
+		ll           *loglist.LogList
+		failGetRoots bool
+		rootNum      map[string]int
+	}{
+		{
+			name:    "InactiveZeroRoots",
+			ll:      sampleValidLogList(t),
+			rootNum: map[string]int{"ct.googleapis.com/aviator/": 0, "ct.googleapis.com/rocketeer/": 2, "ct.googleapis.com/icarus/": 1}, // aviator is not active; 1 of 2 icarus roots is not x509 struct
+		},
+		{
+			name:         "CouldNotCollect",
+			ll:           sampleValidLogList(t),
+			failGetRoots: true,
+			rootNum:      map[string]int{"ct.googleapis.com/aviator/": 0, "ct.googleapis.com/rocketeer/": 2, "ct.googleapis.com/icarus/": 0}, // aviator is not active; icarus client cannot provide roots
+		},
+	}
+
+	for _, tc := range testCases {
+		// Append invalid data to one of Log-stubs that makes GetAcceptedRoots call
+		// return an error.
+		logToFail := "ct.googleapis.com/icarus/"
+		if tc.failGetRoots {
+			RootsCerts[logToFail] = append(RootsCerts[logToFail], "invalidData")
```

Comment:
Instead of complicating the test code, could you not just have this in the test data at the top of the file and use an alternate logList?

Reply:
This particular struct serves get-roots requests. I doubt you can easily bind it into LogList struct.

Reply:
The way I was thinking you'd do this would be to have an entry in `RootsCerts` for "foo-log" with "invalidData" and then the "CouldNotCollect" test could use a logList containing "foo-log" instead of `sampleValidLogList()`.

Reply:
Adapted.